Add data conversion functions.

prideout opened the pull request:
Fixes #46.
 
For the integer-based reading function, note that we return cgltf_size values instead of cgltf_int values.  For the rare cases where clients need to read negative values, they should use the float-based reader instead.

The reason we use size_t is that it's what we need internally for sparse support. Most clients will be reading indices anyway when using the integer-based reader function.

jkuhlmann commented:
Looks good!
The CI builds seem to fail on Linux and Windows though. On Linux, it can't find `std::numeric_limits<float>::lowest()`. Maybe we have to specifically tell it to compile in C++11 mode?
On Windows, it fails when testing `SimpleSparseAccessor.gltf`` as part of the new conversion test executable I think.

zeux commented:
I'm concerned about this interface. This invites the caller to create code that is prone to security vulnerabilities. Here's how I'd use this interface to read position data:

```c++
float pos[3];
cgltf_accessor_read_float(accessor, i, pos);
```

This code is tantalizingly simple, but if input file contains an accessor that's wider or narrower, this either results in stack buffer overrun or uninitialized data. This can be patched by comparing the value type with the expected width, but it's easy to forget. This is part of why I suggested one-at-a-time accessors that would return the value as opposed to filling a buffer.

prideout commented:
Good point @zeux, but to fix this I'd like to simply add an element size argument in [1, 16].

The integer-based function could be one component at a time, but I think the float-based method should return an entire element since that's is almost always what clients want, and it would greatly simplify cases like the three padded matrix types.

I'll look into travis failures too.

prideout commented:
Sorry for the churn guys, but I feel it is ready for another look now.  Here's what I did:
- Added a size argument to the float reader for better security
- Simplified the index reader.
- Made a bug fix for sparse accessors.
- Cleaned up the documentation.
- Fixed Travis.

jkuhlmann commented:
No worries - just making sure everything works as intended. :-)
I'd merge this now or do you have any further input, @zeux ?

zeux commented:
👍

prideout commented:
@jkuhlmann hold off on the merge, I just realized that I'm still not handing sparse accessors incorrectly.

 In fact it's actually not possible to handle them efficiently with a one-element-at-a-time API.  :-\

zeux commented:
I feel like sparse accessors are such a special case that they could be part of a separate API. You can probably build a reverse index map to make single access API practical but it’s definitely not a 90% usecase :)

prideout commented:
For now I removed sparse accessor support so that we can get this in, I do think it would be nice to add another reader function in the future that supports sparse access, since this is an otherwise feature-complete library.
